R/app.R

```diff
@@ -496,3 +506,17 @@ knit_print.reactive <- function(x, ..., inline = FALSE) {
     x()
   }), inline = inline)
 }
+
+#' Run code after app has finished
+#'
+#' This function allows you to register a callback function that is invoked
+#' after your app has finished running (all sessions have been disconnected).
+#'
+#' @param fun A function that will be called after the app has finished running.
+#'
+#' @examples
+#' onStop(function() { print("goodbye") })
+#' @export
+onStop <- function(fun) {
+  shinyOptions(onStop = fun)
```

Comment:
I don't think we want to use this model, where if you call onStop multiple times, the last one wins. Instead we should use a Callback(? I'm on my phone) model so that multiple calls will all be registered and executed, and they can be individually cancelled as well. Like we do for onFlushed. However, one thing I'm not sure we do for onFlushed is execute the callbacks in the reverse order of their registration--this is important for cleanup scenarios (e.g. imagine creating a pool that needs to be cleaned up, and then checking out a connection from that pool; you'd want to clean up the connection and then the pool).